lib/kafkaClient.js

```diff
@@ -999,6 +1004,51 @@ KafkaClient.prototype.verifyPayloadsHasLeaders = function (payloads, callback) {
   });
 };
 
+KafkaClient.prototype.wrapControllerCheckIfNeeded = function (encoder, decoder, encoderArgs, callback) {
+  if (callback.isControllerWrapper) {
+    return callback;
+  }
+
+  var hasBeenInvoked = false;
+
+  const wrappedCallback = (error, result) => {
+    if (error instanceof NotControllerError) {
+      this.setControllerId(null);
+
+      if (!hasBeenInvoked) {
+        hasBeenInvoked = true;
+        this.sendControllerRequest(encoder, decoder, encoderArgs, wrappedCallback);
+        return;
+      }
+    }
+
+    callback(error, result);
+  };
+
+  wrappedCallback.isControllerWrapper = true;
+
+  return wrappedCallback;
+};
+
+KafkaClient.prototype.sendControllerRequest = function (encoder, decoder, encoderArgs, callback) {
+  this.getController((error, controller) => {
+    if (error) {
+      return callback(error);
+    }
+
+    const originalArgs = _.clone(encoderArgs);
+    const originalCallback = callback;
+    const correlationId = this.nextId();
+    encoderArgs.unshift(this.clientId, correlationId);
+    const request = encoder.apply(null, encoderArgs);
+
+    callback = this.wrapControllerCheckIfNeeded(encoder, decoder, originalArgs, originalCallback);
+
+    this.queueCallback(controller.socket, correlationId, [decoder, callback]);
```

Comment:
hi @jlandersen thanks for the update.  Could you help me understand why `wrapControllerCheckIfNeeded` is needed? Would it not be simpler to just  inline a function in place of the `callback` that checks if the error is an instance of `NotControllerError` and then set the controller Id to null and call `sendControllerRequest` again if it is?

Reply:
The reason was to limit the retry to a single time (or potentially x times). If we add a function in place of the callback that always calls the `sendControllerRequest` on a `NotControllerError`, it could keep making requests if the error is returned repeatedly. I am not saying can or will happen, but was rather to be safe. Let me know if you prefer the otherwise :-)